### DISCUSSION TITLE
visible light beam for the Spotlight3D Node

### DISCUSSION THREAD

**Discussion by havi05** [godotengine/godot-proposals#4557] [3D]
I want to create a stage scene with fog and spotlights, but I can't make the beam visible.
I would like to color the parts of the fog (Godot 4.0) that the beam goes through in the color of the spotlights (currently I only see the option to color the entire fog node).
If this is not possible, I would like to have a beam without the fog node.

It would be possible to create a mesh in the shape of a cone, but as the light moves and changes its angle, it would have to be recalculated over and over again. So I think it would be better if it is in the core. (Please tell me if there is an other way.)

I want to create something like this:
![grafik](https://user-images.githubusercontent.com/71942164/168877664-c015adf0-3b63-49ad-ad98-00548bf12a87.png)

**Comment by Calinou** [comment]
This can already be done without requiring real-time mesh generation. There are two ways to achieve this:

- Using a FogVolume node in 4.0alpha. This is more correct (especially when the camera enters the SpotLight's area), but also more expensive to render.
- Using a special billboarded QuadMesh in either 4.0alpha or 3.x. To set the correct billboard orientation, a custom shader is required[^1] (or you need to rotate the plane on the CPU in a script). Enable Proximity Fade on the material to fade the beam when it gets close to solid objects. Also, enable Distance Fade on the material to fade the SpotLight beam when the camera gets too close.
  - If the light's range, spot angle or color changes at run-time, the mesh can be stretched and the material's albedo color can be adjusted.

[^1]: If the SpotLight is pointing exactly upwards or downwards, you do not need a custom shader. Setting the billboard mode to **Y-Billboard** will do the trick :slightly_smiling_face:

**Reply by havi05** [reply]
Thank you, the first point is working really great!

...but I have a Question:
If you have a large FogVolume node, the farther away you are the weaker the light beam becomes. Is there a setting to change how fast the beam view decreases ? (The density setting decreases the fog strengh but also the light intensity)

**Reply by Calinou** [reply]
Increase the **Length** property in the Environment volumetric fog settings:

![image](https://user-images.githubusercontent.com/180032/169159399-6557c1ab-8a23-4207-9fd9-41220fb3809e.png)

Note that doing so will decrease the level of detail of volumetric fog, which can lead to flickering of thin fog volumes. To prevent this (at a performance cost), increase Volume Depth to match the Length property in the Project Settings:

![image](https://user-images.githubusercontent.com/180032/169159344-a7e784fa-9142-49b7-bbc2-c2efd56486ea.png)

**Comment by Calinou** [comment]
I've opened a pull request that adds built-in cone and cylinder shapes to FogVolume: https://github.com/godotengine/godot/pull/61243

This is more suited to SpotLights compared to the existing Ellipsoid shape.
